### DISCUSSION TITLE
Custom events

### DISCUSSION THREAD

**Discussion by ospfranco** [prisma/prisma#11580] [Q&A] [answered]
Hi! I'm trying to use/generate my own custom events from a postgres database, basically I just want to be notified when a field on a table changes, I have written the following trigger on my own migration:

```sql
-- CreateEnum
CREATE TYPE "MessageStatus" AS ENUM ('NEW', 'PROCESSING', 'PROCESSED');

-- CreateTable
CREATE TABLE "MessageQueue" (
    "id" TEXT NOT NULL,
    "queue" TEXT NOT NULL DEFAULT E'default_queue',
    "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updatedAt" TIMESTAMP(3) NOT NULL,
    "status" "MessageStatus" NOT NULL DEFAULT E'NEW',
    "message" JSONB NOT NULL,

    CONSTRAINT "MessageQueue_pkey" PRIMARY KEY ("id")
);

CREATE OR REPLACE FUNCTION message_queue_status_notify()
	RETURNS trigger AS
$$
BEGIN
	PERFORM pg_notify('message_queue_status_channel', NEW.id::text);
	RETURN NEW;
END;
$$ LANGUAGE plpgsql;


CREATE TRIGGER "MessageQueueTrigger"
	AFTER INSERT OR UPDATE OF status
	ON "MessageQueue"
	FOR EACH ROW
EXECUTE PROCEDURE message_queue_status_notify();
```

And then, when I'm initializing the prisma client I want to subscribe to this event channel via the `$on` method:

```ts
import { enhancePrisma } from "blitz"
import { PrismaClient } from "@prisma/client"

const EnhancedPrisma = enhancePrisma(PrismaClient)

export * from "@prisma/client"
const prisma = new EnhancedPrisma()
export default prisma

const isServer = typeof window == "undefined"

if (isServer) {
  prisma.$on("message_queue_status_channel", (event) => {
    console.warn("message queue received new event!", event)
  })
  console.warn("Prisma subscribed to event")
}
```

However, the subscriber callback is not triggered, am I doing something wrong, or is the `$on` method just not meant to support custom events?

**Comment by austincrim** [comment, marked as answer]
Hey @ospfranco 👋,

The [`$on` method](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#on) can only be used with [logging events](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log) or the [`beforeExit` hook](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-management#exit-hooks) which runs before the Prisma client disconnects. 

This is a bit unclear in our docs so I will raise an issue to get that clarified.

**Reply by ospfranco** [reply]
Thanks for the clarification. Somebody let me know on twitter a couple of weeks ago, but it would still be great to have a discussion if custom event support is possible?

**Reply by austincrim** [reply]
It looks like you already opened an issue for this (#11581) which is great! Our engineering teams will take it from there and will ultimately decide when/if we can take this feature on.

Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
